web/src/components/onboarding/OnboardingTaxId.test.tsx

```diff
@@ -139,18 +162,109 @@ describe("<OnboardingTaxId />", () => {
       expect(screen.queryByLabelText("Tax id location")).not.toBeInTheDocument();
     });
 
-    it("clears the full field if user edits the tax id field with * still present", () => {
+    it("renders the hide button initially if the tax id is empty", () => {
       renderComponent(
         {
           ...profileData,
           taxId: "",
         },
         errorProps
       );
-      fireEvent.click(screen.getByLabelText("Tax id"));
-      fireEvent.change(screen.getByLabelText("Tax id"), { target: { value: "******123456" } });
-      fireEvent.blur(screen.getByLabelText("Tax id"));
-      expect((screen.getByLabelText("Tax id") as HTMLInputElement).value).toEqual("");
+      expect(screen.getByText(Config.tax.hideButtonText)).toBeInTheDocument();
+    });
+
+    it("disables the field if the encrypted tax id is populated", () => {
+      renderComponent(
+        {
+          ...profileData,
+          taxId: "********9000",
+          encryptedTaxId: "some-encrypted-value",
+        },
+        errorProps
+      );
+      expect(screen.getByLabelText("Tax id")).toBeDisabled();
+    });
+
+    it("decrypts the tax id if needed when show is clicked and button toggles text to hide", async () => {
+      mockApi.decryptTaxId.mockResolvedValue("123456789000");
+      renderComponent(
+        {
+          ...profileData,
+          taxId: "********9000",
+          encryptedTaxId: "some-encrypted-value",
+        },
+        errorProps
+      );
+      expect(screen.queryByText(Config.tax.hideButtonText)).not.toBeInTheDocument();
+      fireEvent.click(screen.getByText(Config.tax.showButtonText));
+      expect(mockApi.decryptTaxId).toHaveBeenCalled();
+      await waitFor(() => {
+        expect((screen.getByLabelText("Tax id") as HTMLInputElement).value).toEqual("123-456-789/000");
+      });
+      expect(screen.getByText(Config.tax.hideButtonText)).toBeInTheDocument();
+    });
+
+    it("doesn't decrypt the tax id if tax id is already decrypted", async () => {
+      mockApi.decryptTaxId.mockResolvedValue("123456789000");
+      renderComponent(
+        {
+          ...profileData,
+          taxId: "*******89000",
+          encryptedTaxId: "some-encrypted-value",
+        },
+        errorProps
+      );
+      fireEvent.click(screen.getByText(Config.tax.showButtonText));
+      await waitFor(() => {
+        expect((screen.getByLabelText("Tax id") as HTMLInputElement).value).toEqual("123-456-789/000");
+      });
+      fireEvent.click(screen.getByText(Config.tax.hideButtonText));
+      await waitFor(() => {
+        expect(currentProfileData().taxId).toEqual("123456789000");
+      });
+      fireEvent.click(screen.getByText(Config.tax.showButtonText));
+      await waitFor(() => {
+        expect(mockApi.decryptTaxId).toHaveBeenCalledTimes(1);
+      });
```

Comment:
does this need a waitFor if it's not async? I'd that only the first button click needs a waitFor since it's the only one making an API call